multisig/src/lib.rs

```diff
@@ -0,0 +1,348 @@
+    Transfer {
+        receiver_id: AccountId,
+        amount: U128,
+    },
+    AddKey {
+        public_key: Base58PublicKey,
+    },
+    DeleteKey {
+        public_key: Base58PublicKey,
+    },
+    FunctionCall {
+        contract_id: AccountId,
+        method_name: String,
+        args: Base64VecU8,
+        deposit: U128,
+        gas: U64,
+    },
+    SetNumConfirmations {
+        num_confirmations: u32,
+    },
+}
+
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct MultiSigContract {
+    num_confirmations: u32,
+    request_nonce: RequestId,
+    requests: Map<RequestId, MultiSigRequest>,
+    confirmations: Map<RequestId, HashSet<PublicKey>>,
+}
+
+impl Default for MultiSigContract {
+    fn default() -> Self {
+        env::panic(b"Multisig contract should be initialized before usage")
+    }
+}
+
+#[near_bindgen]
+impl MultiSigContract {
+    /// Initialize multisig contract.
+    /// @params num_confirmations: k of n signatures required to perform operations.
+    #[init]
+    pub fn new(num_confirmations: u32) -> Self {
+        assert!(!env::state_exists(), "Already initialized");
+        Self {
+            num_confirmations,
+            request_nonce: 0,
+            requests: Default::default(),
+            confirmations: Default::default(),
+        }
+    }
+
+    /// Add request for multisig.
+    pub fn add_request(&mut self, request: MultiSigRequest) -> RequestId {
+        assert_eq!(
+            env::current_account_id(),
+            env::predecessor_account_id(),
+            "Predecessor account must much current account"
+        );
+        self.requests.insert(&self.request_nonce, &request);
+        let confirmations = HashSet::new();
+        self.confirmations
+            .insert(&self.request_nonce, &confirmations);
+        self.request_nonce += 1;
+        self.request_nonce - 1
+    }
+
+    /// Remove given request and associated confirmations.
+    pub fn delete_request(&mut self, request_id: RequestId) {
+        assert_eq!(
+            env::current_account_id(),
+            env::predecessor_account_id(),
+            "Predecessor account must much current account"
+        );
+        assert!(
+            self.requests.get(&request_id).is_some(),
+            "No such request: either wrong number or already confirmed"
+        );
+        assert!(
+            self.confirmations.get(&request_id).is_some(),
+            "Internal error: confirmations mismatch requests"
+        );
+        self.requests.remove(&request_id);
+        self.confirmations.remove(&request_id);
+    }
+
+    fn execute_request(&mut self, request: MultiSigRequest) -> PromiseOrValue<bool> {
+        match request {
+            MultiSigRequest::Transfer {
+                receiver_id,
+                amount,
+            } => PromiseOrValue::Promise(Promise::new(receiver_id).transfer(amount.into())),
+            MultiSigRequest::AddKey { public_key } => PromiseOrValue::Promise(
```

Comment:
Instead of `PromiseOrValue::Promise(...)` you can use `Promise::new(...).into()`